frame/support/src/storage/stream_iter.rs

```diff
@@ -0,0 +1,405 @@
+		}
+	}
+}
+
+/// An iterator that streams values directly from storage.
+///
+/// Requires that `T` implements the sealed trait `StreamIter`.
+///
+/// Instead of loading the entire `T` into memory, the iterator only loads a certain number of bytes
+/// into memory to decode the next `T::Item`. The iterator implementation is allowed to have some
+/// internal buffer to reduce the number of storage reads. The iterator should have an almost
+/// constant memory usage over its lifetime. If at some point there is a decoding error, the
+/// iterator will return `None` to signal that the iterator is finished.
+pub trait StorageStreamIter<T: private::StreamIter> {
+	/// Create the streaming iterator.
+	fn stream_iter() -> T::Iterator;
+}
+
+impl<T: private::StreamIter + codec::FullCodec, StorageValue: super::StorageValue<T>>
+	StorageStreamIter<T> for StorageValue
+{
+	fn stream_iter() -> T::Iterator {
+		T::stream_iter(Self::hashed_key().into())
+	}
+}
+
+/// A streaming iterator implementation for SCALE container types.
+///
+/// SCALE container types follow the same type of encoding `Compact<u32>(len) ++ data`.
+/// This type provides an [`Iterator`](sp_std::iter::Iterator) implementation that decodes
+/// one item after another with each call to [`next`](Self::next). The bytes representing
+/// the container are also not read at once into memory and instead being read in chunks. As long
+/// as individual items are smaller than these chunks the memory usage of this iterator should
+/// be constant. On decoding errors [`next`](Self::next) will return `None` to signal that the
+/// iterator is finished.
+pub struct ScaleContainerStreamIter<T> {
+	marker: sp_std::marker::PhantomData<T>,
+	input: StorageInput,
+	length: u32,
+	read: u32,
+}
+
+impl<T> ScaleContainerStreamIter<T> {
+	/// Creates a new instance of the stream iterator.
+	///
+	/// - `key`: Storage key of the container in the state.
+	pub fn new(key: Vec<u8>) -> Self {
+		let mut input = StorageInput::new(key);
+		let length = if input.exists() {
+			match codec::Compact::<u32>::decode(&mut input) {
+				Ok(length) => length.0,
+				Err(e) => {
+					// TODO #3700: error should be handleable.
+					log::error!(
+						target: "runtime::storage",
+						"Corrupted state at `{:?}`: {:?}",
+						input.key,
+						e,
+					);
+
+					0
+				},
+			}
+		} else {
+			0
+		};
+
+		Self { marker: sp_std::marker::PhantomData, input, length, read: 0 }
+	}
+}
+
+impl<T: codec::Decode> sp_std::iter::Iterator for ScaleContainerStreamIter<T> {
+	type Item = T;
+
+	fn next(&mut self) -> Option<T> {
+		if self.read >= self.length {
+			None
+		} else {
+			self.read += 1;
+
+			codec::Decode::decode(&mut self.input).ok()
+		}
+	}
+}
+
+/// The size of the internal buffer used by [`StorageInput`].
+///
+/// This internal buffer is used to speed up implementation as reading from the
+/// state for every access is too slow.
+const STORAGE_INPUT_BUFFER_CAPACITY: usize = 2 * 1024;
```

Comment:
Since the buffer's allocated dynamically anyway it doesn't necessarily have to be passed through const generics. (: 

But yeah, it might be worth it to make this adjustable. (Not necessarily now though; as a first attempt I think this is fine as-is, and we can do performance tuning later once this is used in a few places and we know what we have to benchmark.)

Reply:
> (Not necessarily now though; as a first attempt I think this is fine as-is, and we can do performance tuning later once this is used in a few places and we know what we have to benchmark.)

That was sort of my thinking as well.